add test for oss-fuzz

sebpop opened the pull request:
Move the fuzz checker code upstream as requested in this review https://github.com/google/oss-fuzz/pull/1694

Commit:
add test for oss-fuzz

mtl1979 commented on Makefile.in:
```diff
@@ -85,9 +85,18 @@ OBJS = $(OBJC)
+# Run e.g. "make all LIB_FUZZING_ENGINE=/path/to/libFuzzer.a"
+# to link the fuzzer(s) against a real fuzzing engine.
+#
+# OSS-Fuzz will define it's own value for LIB_FUZZING_ENGINE.
```
Should be "its", not "it's"

sebpop replied:
Right.  I copied that comment from the example at https://github.com/google/oss-fuzz/blob/master/projects/example/my-api-repo/Makefile
I will fix it in my patch and send out a patch to fix the example.

mtl1979 commented on test/fuzz/compress_fuzzer.c:
```diff
@@ -0,0 +1,19 @@
+#include "zlib-ng.h"
+
+static Bytef cbuf[256 * 1024] = { 0 };
+static Bytef ubuf[256 * 1024] = { 0 };
```
"Bytef" is discouraged in new code, should use standard types instead.

mtl1979 commented on test/fuzz/compress_fuzzer.c:
```diff
@@ -0,0 +1,19 @@
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  uLongf cbuf_len = (uLongf)sizeof(cbuf);
+  uLongf ubuf_len = (uLongf)sizeof(ubuf);
```
"uLongf" is discouraged in new code, should use standard types instead.

mtl1979 commented on test/fuzz/standalone_fuzz_target_runner.c:
```diff
@@ -0,0 +1,29 @@
+    free(buf);
+    fprintf(stderr, "Done:    %s: (%zd bytes)\n", argv[i], n_read);
+  }
+}
```
Missing return statement.

Commit:
make LLVMFuzzerTestOneInput return 0

Commit:
compress_fuzzer: increase compression buffer size

the asserts in compress_fuzzer may trigger if the output of compress is larger
than the input.

Dead2 commented:
Currently this fuzzer only tests with the default compression level (6), meaning either deflate_medium or _slow, depending on compilation choices. Maybe it would be better to get each of the compression algorithms tested equally? That would require changing to using the compress2 function, and selecting a compression level.

Have a look at configuration_table in deflate.c for example. We have deflate_quick, _fast, _medium and _slow being used at different compression levels. 

How about the fuzzer takes a buffer and compresses it with levels 1, 3, 6, 7 and verifies each of the returned compressed buffers. (lvl 6 could be changed to lvl 4 to increase speed, but since lvl 6 is the default level, that might be an even more compelling reason not to.)


Btw, from a security standpoint, testing broken compressed streams through the decompressor would likely be very valuable. This is where a corrupt .png/.pdf etc file would be able to cause havoc on the unsuspecting users machine.

sebpop commented:
These are very good points: I will add code to the compress_fuzzer to check compression at different levels. 
I will add a decompress_fuzzer that starts decompressing the inputs of the fuzzer, then it will compress that and finally match that the output of decompress+compress matches the original input.

Commit:
compress_fuzzer: check different compression levels

Commit:
compress_fuzzer: add decompress checker

sebpop commented:
> then it will compress that and finally match that the output of decompress+compress matches the original input.

This last part for the decompress checker does not seem to be feasible: there is no unique compression form.

Commit:
compress_fuzzer: add decompress checkers with constrained output size

Commit:
fix bug #184: bzero buf to avoid use-of-uninitialized-value

Commit:
fix bug #183: use memmove to avoid AddressSanitizer: memcpy-param-overlap warnings

Commit:
partial fix for bug #187: move decrement in loop

following bugs remain:
/src/zlib-ng/arch/x86/deflate_quick.c:233:22: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'unsigned int'
/src/zlib-ng/arch/x86/fill_window_sse.c:52:28: runtime error: unsigned integer overflow: 1 - 8192 cannot be represented in type 'unsigned int'

Commit:
fix bug #187: as recommended by Mika Lindqvist

Dead2 commented:
The decompression tester might need a semi-valid header to actually get through to the decompression code more often, and not just fail during the init code (I did not verify that it does bail out during init, but I suspect it might).
Perhaps by testing both without a header and with a prepended header that just specifies the most common mode of using deflate with no preset dictionary.

Btw, maybe the bugfixes should go in a separate PR, so we don't clutter up this PR with comments regarding those.

Commit:
fix bug #190: add a semi-valid gzip header for decompression

Commit:
address review comments

https://github.com/sebpop/zlib-ng/commit/5b5860b7092dc084e15884bc357c9253c72ca3c0#commitcomment-30131656
> Compression level and window size are not really needed for decompression code,
> IIRC they are left there as metadata for information purposes only, so that a
> quick inspection of a file/stream can see whether it might be a good idea to
> re-compress it with a higher compression level.  Because of that, I doubt there
> is much/any gain by running the decompression code for each of those
> combinations. I suspect it will only be a waste of cpu time.

To speed up the fuzzer code, remove memcpy of the input data, and call
check_decompress once with no preset dictionary.

Commit:
[NFC] rename variables to match test/example.c

No functionality change.

Commit:
[NFC] make data and dataLen global static variables

Commit:
add all test_* functions from test/example.c

Commit:
fix bug #183 following recommendations of Mika Lindqvist

> the problem is in line c_stream.avail_in = (unsigned int)comprLen/2;
> which feeds it too much data ... it should cap it to
> c_stream.next_out - compr instead.

Commit:
add fuzzer to make check

Commit:
fix use-of-uninitialized-value

==397==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4e8193 in fill_window_sse /src/zlib-ng/arch/x86/fill_window_sse.c:52:30
    #1 0x4e9254 in deflate_quick /src/zlib-ng/arch/x86/deflate_quick.c:220:13
    #2 0x4ad375 in zng_deflate /src/zlib-ng/deflate.c:951:18
    #3 0x49fef1 in test_large_deflate /src/zlib-ng/test/example.c:264:11
    #4 0x4a3232 in main /src/zlib-ng/test/example.c:536:5
    #5 0x7f927f42d82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #6 0x41b3a8 in _start (/src/zlib-ng/example+0x41b3a8)
  Uninitialized value was created by a heap allocation
    #0 0x45b9e0 in malloc /src/llvm/projects/compiler-rt/lib/msan/msan_interceptors.cc:910
    #1 0x4a47b2 in zng_deflateInit2_ /src/zlib-ng/deflate.c:284:27
    #2 0x4a3f6a in zng_deflateInit_ /src/zlib-ng/deflate.c:224:12
    #3 0x49fdd5 in test_large_deflate /src/zlib-ng/test/example.c:253:11
    #4 0x4a3232 in main /src/zlib-ng/test/example.c:536:5
    #5 0x7f927f42d82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Commit:
fix bug #194: only run test_flush when input stream has more than 3 bytes

Commit:
pass to compress_fuzzer all *.* existing files

some testers were failing to find compress.c as they were using
a separate build dir. To avoid this and to expand the coverage of the test
we now pass all *.* files to compress_fuzzer.

Commit:
make abort error messages unique

Commit:
stand-alone-fuzzer: close file stream once finished

Commit:
compress_fuzzer: clear out allocated data

Commit:
compress_fuzzer: reduce size of dictionary when dataLen is smaller than the dictionary

Commit:
make abort error messages unique

Commit:
disable large deflate/inflate tests until they get fixed

Commit:
add cmake test for compress_fuzzer

Commit:
make: use wildcard to match files

Commit:
make: use bash for compress_fuzzer

Commit:
make: use QEMU_RUN to run compress_fuzzer

Commit:
increase compr/uncompr buffer sizes

to avoid the following two errors:

Running: libz.a
deflate not greedy

Running: deflate_quick.lo
deflate large should report Z_STREAM_END

Commit:
fix fallout from bug #184: do not use bzero

Mika Lindqvist suggested:
> You shouldn't use bzero() in new code as some compilers, like Visual C++,
> don't have it... New code should just use memset().

Commit:
fix AddressSanitizer: heap-buffer-overflow

Commit:
fix MemorySanitizer: use-of-uninitialized-value

Commit:
fix bugs #186 and #191: use-of-uninitialized-value

==1==WARNING: MemorySanitizer: use-of-uninitialized-value
  #0 0x59fa93 in deflate_medium zlib-ng/deflate_medium.c:259:21
  #1 0x590905 in zng_deflate zlib-ng/deflate.c:951:18
  #2 0x587095 in zng_compress2 zlib-ng/compress.c:59:15
  #3 0x5866e3 in check_compress_level zlib-ng/test/fuzz/compress_fuzzer.c:18:3
  #4 0x5862fd in LLVMFuzzerTestOneInput zlib-ng/test/fuzz/compress_fuzzer.c:38:3
  #5 0x4e9b48 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:575:15
  #6 0x4a2f66 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:280:6
  #7 0x4b3adb in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:715:9
  #8 0x4a2091 in main /src/libfuzzer/FuzzerMain.cpp:20:10
  #9 0x7fea2fea482f in __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/libc-start.c:291
  #10 0x41ec68 in _start
Uninitialized value was created by a heap allocation
  #0 0x45f2a0 in malloc /src/llvm/projects/compiler-rt/lib/msan/msan_interceptors.cc:910
  #1 0x587d42 in zng_deflateInit2_ zlib-ng/deflate.c:284:27
  #2 0x5874fa in zng_deflateInit_ zlib-ng/deflate.c:224:12
  #3 0x586c95 in zng_compress2 zlib-ng/compress.c:41:11
  #4 0x5866e3 in check_compress_level zlib-ng/test/fuzz/compress_fuzzer.c:18:3
  #5 0x5862fd in LLVMFuzzerTestOneInput zlib-ng/test/fuzz/compress_fuzzer.c:38:3
  #6 0x4e9b48 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:575:15
  #7 0x4a2f66 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:280:6
  #8 0x4b3adb in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:715:9
  #9 0x4a2091 in main /src/libfuzzer/FuzzerMain.cpp:20:10
  #10 0x7fea2fea482f in __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/libc-start.c:291

Commit:
do not overwrite input of fuzzer

to avoid this error:
==12== ERROR: libFuzzer: fuzz target overwrites it's const input

Commit:
fix bug #192: MemorySanitizer:DEADLYSIGNAL

==4908==ERROR: MemorySanitizer: SEGV on unknown address 0x730fffffffff (pc 0x0000004b1b97 bp 0x7ffd4bf59a00 sp 0x7ffd4bf598a0 T4908)
==4908==The signal is caused by a READ memory access.

Commit:
[fuzz] split out compress_fuzzer into 5 smaller tests

ninja test shows the following total running times:

2/7 Test #2: compress_fuzzer ..................   Passed    0.67 sec
3/7 Test #3: example_small_fuzzer .............   Passed    0.55 sec
4/7 Test #4: example_large_fuzzer .............   Passed    0.12 sec
5/7 Test #5: example_flush_fuzzer .............   Passed    0.20 sec
6/7 Test #6: example_dict_fuzzer ..............   Passed    1.16 sec

The patch also restricts all the fuzzers to process less than 1Mb input.
We will probably need to restrict the dict_fuzzer to less than 1Mb.

Commit:
[fuzz] reduce input size

Commit:
compress_fuzzer: fix memory leak

Commit:
restrict example_dict_fuzzer to 100Kb to avoid timeouts

Commit:
fix #197, oss-fuzz/10036: only write 4 bytes per iteration in deflate_quick

by aggregating the two consecutive values to be written by static_emit_ptr to
s->pending_buf and writing the two values at once in a 4 byte store, we avoid
running out of the allocated buffer. We used to call quick_send_bits twice and
bumped the counter s->pending in the first call, which made the second call
write to memory beyond the safe 4 bytes that were guaranteed by the following
condition in the enclosing loop in deflate_quick:

  if (s->pending + 4 >= s->pending_buf_size) {
    flush_pending(s->strm);

The bug was exposed by the memory sanitizer like so:

MemorySanitizer:DEADLYSIGNAL
--
  | ==1==ERROR: MemorySanitizer: SEGV on unknown address 0x730000020000 (pc 0x0000005b6ce4 bp 0x7fff59adb5e0 sp 0x7fff59adb570 T1)
  | ==1==The signal is caused by a WRITE memory access.
  | #0 0x5b6ce3 in quick_send_bits zlib-ng/arch/x86/deflate_quick.c:134:48
  | #1 0x5b5752 in deflate_quick zlib-ng/arch/x86/deflate_quick.c:243:21
  | #2 0x590a15 in zng_deflate zlib-ng/deflate.c:952:18
  | #3 0x587165 in zng_compress2 zlib-ng/compress.c:59:15
  | #4 0x5866d3 in check_compress_level zlib-ng/test/fuzz/compress_fuzzer.c:22:3
  | #5 0x5862d8 in LLVMFuzzerTestOneInput zlib-ng/test/fuzz/compress_fuzzer.c:74:3
  | #6 0x4e9b48 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:575:15
  | #7 0x4a2f66 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:280:6
  | #8 0x4b3adb in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:715:9
  | #9 0x4a2091 in main /src/libfuzzer/FuzzerMain.cpp:20:10
  | #10 0x7fb8919b082f in __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/libc-start.c:291
  | #11 0x41ec68 in _start
  | MemorySanitizer can not provide additional info.
  | SUMMARY: MemorySanitizer: SEGV (/mnt/scratch0/clusterfuzz/slave-bot/builds/clusterfuzz-builds_zlib-ng_7ead0a3e4980f024583384fd355b6e3ddd4b2ca2/revisions/compress_fuzzer+0x5b6ce3)

sebpop commented:
Now that the oss-fuzzers are free of bugs and that the patches have stabilized, I think it is time to think about merging the patch series.
@Dead2 : could you please let me know how you would like to proceed?
I can rework the patches following your guidelines and create separate pull requests for each fixed bug.

Dead2 commented:
@sebpop Separate pull requests for each fixed bug is great, just like I see you have started doing. 👍 
Also a separate PR for the oss-fuzz support would be great.
I am a bit unsure about the attribution requirements from oss-fuzz though, didn't they require some text in the commit message?

@mtl1979 Please +1 the PRs you approve of if you can, I value your feedback.
Any others doing the same would be great. The more eyes reviewing zlib-ng changes, the better both for me and for the projects potential inclusion in distros.  :)

@sebpop Great work on this! If you ever visit (east) Norway, I'll buy you a beer or something :)

mtl1979 commented:
@Dead2 Already started going through the ones that seem ready for merging... Some of the changes did seem still work-in-progress so I'm waiting for @sebpop to finish them. 

I'm currently quite busy with work so I can't go through the modifications every day, but I try to go through them when I have spare time.

Dead2 commented:
@mtl1979 Great, I am not rushing anything here. But knowing you intend to review, I'll hold off on merging until I hear your feedback on them. Oh, and that beer offer goes for you too :)

mtl1979 commented:
@Dead2 Everybody knows I don't drink beer...

Dead2 commented:
@mtl1979 Me neither ;)

mtl1979 commented:
@Dead2 I did empty my last bottle of alcohol about 1 year after I stopped working for local recycling center... After that I have not even tasted alcohol.

sebpop commented:
> Also a separate PR for the oss-fuzz support would be great.

I just opened bug #204.

> I am a bit unsure about the attribution requirements from oss-fuzz though, didn't they require some text in the commit message?

For the bugs with an oss-fuzz ticket, I added that reference in the subject of the patch fixing it.
If you want I can also add the reference in the commit message.

> the projects potential inclusion in distros

I'd like to get zlib-ng in chromium.org and Android.  oss-fuzzer is going to help stabilize the code, and the big performance difference between zlib-ng and zlib.org already provides the incentive to make the switch.

> If you ever visit (east) Norway, I'll buy you a beer or something :)

I like to keep my brain clear to work on code, I haven't got a beer from the time I was in Liverpool.
Hot chocolate or a tea would work ;-)

Commit:
fix oss-fuzz/10070: be more careful with shifts

to avoid the following ubsan error:
/src/zlib-ng/arch/x86/deflate_quick.c:134:66: runtime error: left shift of 2049 by 20 places cannot be represented in type 'int'
#0 0x4759d4 in quick_send_bits zlib-ng/arch/x86/deflate_quick.c:134:66
#1 0x474d7a in deflate_quick zlib-ng/arch/x86/deflate_quick.c:246:21
#2 0x46683c in zng_deflate zlib-ng/deflate.c:952:18
#3 0x4635b5 in zng_compress2 zlib-ng/compress.c:59:15
#4 0x46325c in check_compress_level zlib-ng/test/fuzz/compress_fuzzer.c:22:3
#5 0x463168 in LLVMFuzzerTestOneInput zlib-ng/test/fuzz/compress_fuzzer.c:74:3
#6 0x44015c in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:575:15
#7 0x430e9d in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:280:6
#8 0x434dbb in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:715:9
#9 0x430b88 in main /src/libfuzzer/FuzzerMain.cpp:20:10
#10 0x7f4cacc7c82f in __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/libc-start.c:291
#11 0x405cb8 in _start

Commit:
compile and run *_fuzzers with ASan/UBSan/MSan

Commit:
configure: pass CFLAGS to CC

when cross compiling with clang, one sets -target aarch64-linux-gnu in the
CFLAGS. Without this patch clang is invoked without target and sets GCC_ARCH
to the default target.

Commit:
[fuzzer] reduce size of dictionary to speed up fuzzers

Commit:
[fuzzer] read the length of the dictionary from input data

Commit:
fuzzers: use compressBound and deflateBound

Commit:
fuzzers: call deflateInit2 with parameters from data[0]

to increase coverage, use the data from the fuzzers to initialize the
compression stream.

Commit:
oss-fuzz/10388: remove strcpy calls from fuzzers

we now allocate the memory based on compressBound which is stricter than the previous computation

Commit:
oss-fuzz/10388: do not call compressBound in example_large_fuzzer

the code in example_large_fuzzer needs more space than what is allocated by compressBound.
To make sure this oss-fuzz issue is fixed, I have tested this patch with the following script:

 #!/bin/bash -eux
cd ~/oss-fuzz
python infra/helper.py build_image zlib-ng
for san in address undefined memory; do
    python infra/helper.py build_fuzzers --clean --sanitizer $san --engine libfuzzer zlib-ng ~/zlib-ng/
    for fuzzer in compress_fuzzer example_small_fuzzer example_large_fuzzer example_flush_fuzzer example_dict_fuzzer; do
        python infra/helper.py check_build --sanitizer $san --engine libfuzzer zlib-ng $fuzzer
    done
done

sebpop commented:
I am closing this old pull request.  The squashed patch to add the fuzzers in a single patch is #206.
